xfer/buffer.go

```diff
@@ -2,40 +2,73 @@ package xfer
 
 import (
 	"bytes"
+	"io"
 	"sync"
 	"sync/atomic"
 )
 
 // A Buffer is a reference counted bytes.Buffer, which belongs
 // to a sync.Pool
-type Buffer struct {
+type Buffer interface {
+	io.Reader
+	io.Writer
+
+	// Get returns a new buffer sharing the contents of this
+	// buffer, but with its own cursor.  Get also increases
+	// the reference count.  It is safe for concurrent calls.
+	Get() Buffer
+
+	// Put decreases the reference count, and when it hits zero, puts the
+	// buffer back in the pool.
+	Put()
+}
+
+type baseBuffer struct {
 	bytes.Buffer
 	pool *sync.Pool
 	refs int32
 }
 
+type dependantBuffer struct {
+	*bytes.Buffer
+	buf *baseBuffer
+}
+
 // NewBuffer creates a new buffer
-func NewBuffer(pool *sync.Pool) *Buffer {
-	return &Buffer{
+func NewBuffer(pool *sync.Pool) Buffer {
+	return &baseBuffer{
 		pool: pool,
 		refs: 0,
 	}
 }
 
-// Get increases the reference count.  It is safe for concurrent calls.
-func (b *Buffer) Get() {
+// Get implements Buffer
```

Comment:
I must be missing something.  For *baseBuffer to implement buffer, we only need to implement Get and Put.  The code wouldn't compile if this wasn't the case.